gstd/src/builder.rs

```diff
@@ -0,0 +1,870 @@
+/// such as `msg::send_bytes()`, `msg::reply_bytes()`.
+pub type BytesMessageBuilder<'a, Buffer> = MessageBuilder<'a, Buffer, (), RangeFull>;
+
+/// Create a new [`BytesMessageBuilder`].
+pub const fn bytes<'a, Buffer: AsRef<[u8]>>() -> BytesMessageBuilder<'a, Buffer> {
+    BytesMessageBuilder::new()
+}
+
+/// This data type provides a more convenient wrapper over functions
+/// such as `msg::send<E: Encode>()`, `msg::reply<E: Encode>()`.
+pub type EncodeMessageBuilder<'a, Encodable> = MessageBuilder<'a, [u8; 0], Encodable, RangeFull>;
+
+/// Create a new [`EncodeMessageBuilder`].
+pub const fn encode<'a, Encodable: Encode>() -> EncodeMessageBuilder<'a, Encodable> {
+    EncodeMessageBuilder::new()
+}
+
+/// This data type provides a more convenient wrapper over functions
+/// such as `msg::send_input<R: RangeBounds<usize>>()`, `msg::reply_input<R>()`.
+pub type InputMessageBuilder<'a, Range> = MessageBuilder<'a, [u8; 0], (), Range>;
+
+/// Create a new [`InputMessageBuilder`].
+pub const fn input<'a, Range: RangeBounds<usize> + Copy>() -> InputMessageBuilder<'a, Range> {
+    InputMessageBuilder::new()
+}
+
+/// Provides an alternative way to interact with the messages API.
+///
+/// Unlike the traditional way of using imperative functions,
+/// it allows calls to be made through a fluent methods.
+///
+/// Instead of writing imperative code like this:
+/// ```no_run
+/// use gstd::{msg, ActorId};
+///
+/// msg::send_bytes_with_gas_delayed(ActorId::zero(), b"PING", 1_000_000, 0, 60)
+///     .expect("failed to send msg");
+/// ```
+///
+/// This wrapper allows you to write code like this:
+/// ```no_run
+/// use gstd::{builder, ActorId};
+///
+/// builder::bytes::<_>()
+///     .to(ActorId::zero())
+///     .payload_bytes(b"PING")
+///     .value(0)
+///     .with_gas(1_000_000)
+///     .delayed(60)
+///     .send()
+///     .expect("failed to send msg");
+/// ```
+pub struct MessageBuilder<'a, Buffer, Encodable, Range>
+where
+    Buffer: AsRef<[u8]>,
+    Encodable: Encode,
+    Range: RangeBounds<usize> + Copy,
+{
+    code_id: Option<CodeId>,
+    program: Option<ActorId>,
+    payload: Option<MessagePayload<'a, Buffer, Encodable, Range>>,
+    value: Option<u128>,
+    delay: Option<u32>,
+    gas_limit: Option<u64>,
+    reservation_id: Option<ReservationId>,
+}
+
+impl<'a, Buffer, Encodable, Range> MessageBuilder<'a, Buffer, Encodable, Range>
+where
+    Buffer: AsRef<[u8]>,
+    Encodable: Encode,
+    Range: RangeBounds<usize> + Copy,
+{
+    /// Create a new [`MessageBuilder`] with zeroed fields.
+    pub const fn new() -> Self {
+        Self {
+            code_id: None,
+            program: None,
+            payload: None,
+            value: None,
+            delay: None,
+            gas_limit: None,
+            reservation_id: None,
+        }
+    }
+
+    /// Sets `code_id` which can be used to call
+    /// [`create_program`](Self::create_program) later.
+    pub const fn code_id(mut self, code_id: CodeId) -> Self {
```

Comment:
I think it would be redundant because we have different setters like
- `pub const fn payload_encode(mut self, payload: &'a Encodable) -> Self` (by ref)
- `pub const fn payload_input(mut self, payload: Range) -> Self` (by value)

Also note that they are marked `const fn` to achieve zero-cost abstraction. Implementation takes much more lines of code than these setters